### PR TITLE
vCenter monitoring

### PR DESCRIPTION
**Pull request by rk4n**
Added implementation of LM monitoring for VMWare vCenter.

### CODE REVIEW

File: migrationmonitor/common/utils.py [hunk 1]
```diff
@@ -0,0 +1,37 @@
+from time import sleep
+
+try:
+    xrange
+except NameError:
+    xrange = range
+
+def retry(max_tries, delay=1, delay_mult=2, exceptions=(Exception,), hook=None):
```
Comment (vladiskuz) [inline migrationmonitor/common/utils.py]: I guess that we can use retry_call() from the retry library instead this decorator https://pypi.python.org/pypi/retry/

File: migrationmonitor/libvirt/libvirt_utils.py [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+import libvirt
+
+
+def run_native_event_loop():
+    """Runs native libvirt event loop
+    """
+    while True:
+        libvirt.virEventRunDefaultImpl()
+
+EVENT_LOOP_THREAD = None
```
Comment (vladiskuz) [inline migrationmonitor/libvirt/libvirt_utils.py]: Let's use words in all capital letters for constants https://www.python.org/dev/peps/pep-0008/#constants

File: migrationmonitor/libvirt/monitor.py [hunk 3]
```diff
@@ -2,21 +2,21 @@
...
+from migrationmonitor.common.db import InfluxDBActor
+from migrationmonitor.libvirt.watcher import LibvirtDomainsWatcher
 
 EVENT_DETAILS = (
     ("Added", "Updated"),
     ("Removed",),
     ("Booted", "Migrated", "Restored", "Snapshot", "Wakeup"),
-    ("Paused", "Migrated", "IOError", "Watchdog",
-        "Restored", "Snapshot", "API error"),
+    ("Paused", "Migrated", "IOError", "Watchdog", "Restored", "Snapshot", "API error"),
```
Comment (vladiskuz) [inline migrationmonitor/libvirt/monitor.py]: This line longer than 79 character https://www.python.org/dev/peps/pep-0008/#maximum-line-length

File: migrationmonitor/libvirt/monitor.py [hunk 4]
```diff
@@ -2,21 +2,21 @@
...
     ("Unpaused", "Migrated", "Snapshot"),
-    ("Shutdown", "Destroyed", "Crashed",
-        "Migrated", "Saved", "Failed", "Snapshot"),
+    ("Shutdown", "Destroyed", "Crashed", "Migrated", "Saved", "Failed", "Snapshot"),
```
Comment (vladiskuz) [inline migrationmonitor/libvirt/monitor.py]: The same as the previous comment

File: migrationmonitor/libvirt/monitor.py [hunk 5]
```diff
@@ -72,8 +76,8 @@ def _register_libvirt_callbacks(self, conn):
             None)
         conn.setKeepAlive(5, 3)
 
+
```
Comment (vladiskuz) [inline migrationmonitor/libvirt/monitor.py]: Blank line

File: migrationmonitor/main.py [hunk 6]
```diff
@@ -3,22 +3,22 @@
...
 def main():
-    utils.start_event_loop()
-    libvirt_monitor = monitor.LibvirtMonitor()
-    libvirt_monitor.start()
+    monitor = VCenterMonitor()
+    monitor.start()
 
     old_exitfunc = getattr(sys, 'exitfunc', None)
 
+
```
Comment (vladiskuz) [inline migrationmonitor/main.py]: blank line

File: migrationmonitor/main.py [hunk 7]
```diff
@@ -37,3 +35,4 @@ def exitfunc():
                            action=main,
                            keep_fds=[log.handler.stream.fileno()])
         daemon.start()
+
```
Comment (vladiskuz) [inline migrationmonitor/main.py]: blank line

File: migrationmonitor/main.py [hunk 8]
```diff
@@ -3,22 +3,22 @@
...
-import utils
+from migrationmonitor.common import logger as log
+from migrationmonitor.libvirt.monitor import LibvirtMonitor
+from vcenter.monitor import VCenterMonitor
+import migrationmonitor.settings
 
 
 def main():
-    utils.start_event_loop()
-    libvirt_monitor = monitor.LibvirtMonitor()
```
Comment (vladiskuz) [inline migrationmonitor/main.py]: How will we use Libvirt Monitor? I suggest using an option in the configuration file for this purpose

File: migrationmonitor/libvirt/watcher.py [hunk 9]
```diff
@@ -99,26 +100,25 @@ def __init__(self, conn, dom_id, migration_monitors, db_actor):
...
         try:
             dom = self.conn.lookupByID(self.dom_id)
             job_info = dom.jobStats()
             log.debug("jobStats: {0}".format(job_info))
-            self.db_actor.tell(({
-                "domain_id": self.dom_id,
-                "domain_name": dom.name()},
-                job_info,
-                self.settings.INFLUXDB["JOBINFO_MEASUREMENT"]))
+            self.db_actor.tell(({"domain_id": self.dom_id,
```
Comment (vladiskuz) [inline migrationmonitor/libvirt/watcher.py]: Unnecessary changes that add unnecessary information to the history

File: migrationmonitor/common/__init__.py [hunk 10]
```diff
@@ -0,0 +1 @@
+
```
Comment (vladiskuz) [inline migrationmonitor/common/__init__.py]: blank line

File: migrationmonitor/libvirt/libvirt_utils.py [hunk 11]
```diff
@@ -0,0 +1,27 @@
+import threading
```
Comment (vladiskuz) [inline migrationmonitor/libvirt/libvirt_utils.py]: If it already placed in the libvirt folder I think that we can name this just like utils.py

File: migrationmonitor/vcenter/monitor.py [hunk 12]
```diff
@@ -0,0 +1,141 @@
+import ssl
+import time
+from collections import deque
+from datetime import datetime
+
+from pyVim import connect
+from pyVmomi import vmodl, vim
```
Comment (vladiskuz) [inline migrationmonitor/vcenter/monitor.py]: let's change this two lines to `from pyVim import connect as vmware_connect` and for the second line

File: migrationmonitor/vcenter/monitor.py [hunk 13]
```diff
@@ -0,0 +1,141 @@
...
+                          event)
+            else:
+                log.debug("Event: %s already reported.", event_id)
+
+
+        time.sleep(settings.VCENTER["POLL_FREQ"])
+        self.tell("continue")
+
+
+    def _reconnect(self, tries_remaining, ex, _delay):
```
Comment (vladiskuz) [inline migrationmonitor/vcenter/monitor.py]: It's not so obviously why we need the `tries_remaining`, `ex`, `_delay`

File: migrationmonitor/common/utils.py (hunk 1 above)
Comment (rk4n) [inline migrationmonitor/common/utils.py]: it doesn't support hooks

File: migrationmonitor/vcenter/monitor.py (hunk 13 above)
Comment (rk4n) [inline migrationmonitor/vcenter/monitor.py]: See the retry implementation, you just need to follow this convention for function signature.

File: migrationmonitor/vcenter/monitor.py (hunk 12 above)
Comment (rk4n) [inline migrationmonitor/vcenter/monitor.py]: It should be pretty obvious that its about vmware, since this file is under `vcenter` folder.

File: migrationmonitor/libvirt/watcher.py (hunk 9 above)
Comment (rk4n) [inline migrationmonitor/libvirt/watcher.py]: So, roll them back?

File: migrationmonitor/main.py (hunk 8 above)
Comment (rk4n) [inline migrationmonitor/main.py]: Good point, my bad.

File: migrationmonitor/libvirt/libvirt_utils.py (hunk 11 above)
Comment (rk4n) [inline migrationmonitor/libvirt/libvirt_utils.py]: ++

File: .gitignore [hunk 14]
```diff
@@ -1,4 +1,9 @@
 *.pyc
```
Comment (rk4n) [inline .gitignore]: What for?

File: .editorconfig [hunk 15]
```diff
@@ -0,0 +1,17 @@
+# http://editorconfig.org
```
Comment (rk4n) [inline .editorconfig]: We actually do, this file is recognized by most of modern editors and helps you automatically maintain proper indents.
